### PR TITLE
fix(api): ensure that window functions are propagated

### LINKED ISSUE #4722: bug(duckdb): ProgrammingError when evaluating ibis.coalesce(t.value.lag(), 0)
## Reproducer

Setup:

```python
>>> import ibis
... import pandas as pd
... 
... t = ibis.memtable(
...     pd.DataFrame(
...         {
...             "value": [1, 2, 3, 4, 5],
...             "group": ["a", "b", "b", "b", "a"],
...         }
...     )
... )
```

when executing the following:

```python
>>> (
...     t.group_by("group")
...     .order_by("value")
...     .mutate(lagged_value=ibis.coalesce(t.value.lag(), 0))
... ).execute()
```

one gets the following:

```
ProgrammingError: (duckdb.CatalogException) Catalog Error: Scalar Function with name lag does not exist!
Did you mean "log"?
LINE 1: SELECT t0.value, t0."group", coalesce(lag(t0.value, ?), CAST(? AS SMALLINT)) ...
```

Full traceback:

<details>

```python-traceback
Traceback (most recent call last):
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1900 in _execute_context
    self.dialect.do_execute(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/default.py:736 in do_execute
    cursor.execute(statement, parameters)
  File ~/code/duckdb_engine/duckdb_engine/__init__.py:120 in execute
    self.c.execute(statement, parameters)
CatalogException: Catalog Error: Scalar Function with name lag does not exist!
Did you mean "log"?
LINE 1: SELECT t0.value, t0."group", coalesce(lag(t0.value, ?), CAST(? AS SMALLINT)) ...
                                              ^

The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  Input In [3] in <cell line: 2>
    ).execute()
  File ~/code/ibis/ibis/expr/types/core.py:274 in execute
    return self._find_backend(use_default=True).execute(
  File ~/code/ibis/ibis/backends/base/sql/__init__.py:256 in execute
    with self._safe_raw_sql(sql, **kwargs) as cursor:
  File ~/mambaforge/envs/dev/lib/python3.10/contextlib.py:135 in __enter__
    return next(self.gen)
  File ~/code/ibis/ibis/backends/base/sql/alchemy/__init__.py:118 in _safe_raw_sql
    yield con.execute(*args, **kwargs)
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1380 in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/sql/elements.py:333 in _execute_on_connection
    return connection._execute_clauseelement(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1572 in _execute_clauseelement
    ret = self._execute_context(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1943 in _execute_context
    self._handle_dbapi_exception(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2124 in _handle_dbapi_exception
    util.raise_(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/compat.py:208 in raise_
    raise exception
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1900 in _execute_context
    self.dialect.do_execute(
  File ~/mambaforge/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/default.py:736 in do_execute
    cursor.execute(statement, parameters)
  File ~/code/duckdb_engine/duckdb_engine/__init__.py:120 in execute
    self.c.execute(statement, parameters)
ProgrammingError: (duckdb.CatalogException) Catalog Error: Scalar Function with name lag does not exist!
Did you mean "log"?
LINE 1: SELECT t0.value, t0."group", coalesce(lag(t0.value, ?), CAST(? AS SMALLINT)) ...
                                              ^
[SQL: SELECT t0.value, t0."group", coalesce(lag(t0.value, ?), CAST(? AS SMALLINT)) AS lagged_value 
FROM _ibis_memtable0 AS t0 
 LIMIT ?]
[parameters: (1, 0, 10000)]
(Background on this error at: https://sqlalche.me/e/14/f405)
```

</details>

## Workaround

The following slightly more verbose solution works:

```python
>>> (
...     t.group_by("group")
...     .order_by("value")
...     .mutate(lagged_value=t.value.lag())
...     .mutate(lagged_value=ibis.coalesce(ibis._.lagged_value, 0))
... ).execute()
   value group  lagged_value
0      1     a             0
1      5     a             1
2      2     b             0
3      3     b             2
4      4     b             3
```

### PR DESCRIPTION
**Pull request by cpcloud**
This PR fixes window function propagation.

I had to refactor some of the operations to be variadic, so that I didn't have
to hack in support for `NodeList` inside of `propagate_down_window`.

Resolves #4722.

### CODE REVIEW

File: ibis/backends/dask/execution/reductions.py [hunk 1]
```diff
@@ -43,28 +43,40 @@ def pairwise_reducer(func, values):
...
 
-@execute_node.register(ops.Greatest, collections.abc.Sequence)
-def dask_execute_node_greatest_list(op, value, **kwargs):
-    if all(type(v) != dd.Series for v in value):
-        return execute_node_greatest_list(op, value, **kwargs)
-    return compute_row_reduction(da.maximum, value)
+try:
+    del execute_node[ops.Greatest, Variadic[object]]
+except KeyError:
+    pass
```
Comment (gforsyth) [inline ibis/backends/dask/execution/reductions.py:59]: What on earth is going on here?

Comment (cpcloud) [reply ibis/backends/dask/execution/reductions.py:67]: Added a comment!

**Review by gforsyth — APPROVED** [review]
Nice!

**Comment by ogrisel** [comment]
🙏

**Pull request merged**